components/calendar/src/chinese_based.rs

```diff
@@ -309,6 +337,159 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
+    /// Also returns a ChineseBasedCache - either the cache passed as an argument, if applicable, or a new
+    /// ChineseBasedCache for the given year.
+    fn months_in_year_cached(
+        year: i32,
+        cache_arg: Option<ChineseBasedCache>,
+    ) -> (u8, ChineseBasedCache) {
+        let (is_leap_year, cache) = Self::is_leap_year_cached(year, cache_arg);
+        (if is_leap_year { 13 } else { 12 }, cache)
+    }
+
+    /// Returns true if the given year is a leap year, and false if not; also returns a ChineseBasedCache -
+    /// either the cache passed as an argument, if applicable, or a new ChineseBasedCache for the given year.
+    fn is_leap_year_cached(
+        year: i32,
+        cache_arg: Option<ChineseBasedCache>,
+    ) -> (bool, ChineseBasedCache) {
+        if let Some(existing_cache) = cache_arg {
+            (existing_cache.is_leap_year, existing_cache)
+        } else {
+            let mid_year = Self::fixed_mid_year_from_year(year);
+            Self::fixed_date_is_in_leap_year_empty_cache(mid_year)
+        }
+    }
+
+    /// Returns true if the given fixed date is in a leap year; also returns a reference
+    /// to a new mutable ChineseBasedCache.
+    fn fixed_date_is_in_leap_year_empty_cache(date: RataDie) -> (bool, ChineseBasedCache) {
+        let (new_year, solstice) = Self::new_year_on_or_before_fixed_date(date, None);
+        let is_leap_year = Self::new_year_is_leap_year(new_year, Some(solstice));
+        let cache = ChineseBasedCache {
+            new_year,
+            is_leap_year,
+        };
+        (is_leap_year, cache)
+    }
+
+    /// Gets the leap month for an instance of `ChineseBasedDateInner`; this function assumes the `ChineseBasedDateInner`
+    /// represents a year with a leap month.
+    pub(crate) fn get_leap_month(&self) -> u8 {
+        debug_assert!(
+            self.1.is_leap_year,
+            "fn `get_leap_month` called on a non-leap year!"
+        );
+        let mut cur = self.1.new_year;
+        let mut result = 1;
+        while result < MAX_ITERS_FOR_MONTHS_OF_YEAR && !Self::no_major_solar_term(cur) {
+            cur = Self::new_moon_on_or_after((cur + 1).as_moment());
+            result += 1;
+        }
+        debug_assert!(
+            result < MAX_ITERS_FOR_MONTHS_OF_YEAR,
+            "Unexpected number of iterations occurred searching for a leap month."
+        );
+        result
+    }
+
+    /// Calls `days_in_month_cached` on an instance of ChineseBasedDateInner
+    pub(crate) fn days_in_month_inner(&self) -> u8 {
+        Self::days_in_month_cached(self.0.year, self.0.month, Some(self.1)).0
+    }
+
+    /// Returns the number of days in the given `year`, `month` (see `month_days`), as well as a ChineseBasedCache -
+    /// either a cache passed in optionally as an argument, or a new cache otherwise.
+    fn days_in_month_cached(
+        year: i32,
+        month: u8,
+        cache_arg: Option<ChineseBasedCache>,
```

Comment:
issue: this only ever gets called with Some

Reply:
I've replaced all these with `ChineseBasedCache` instead of `Option`s.